### PR TITLE
Added Inference Option to Tokenizer/Transform

### PR DESCRIPTION
**Pull request by pbontrager**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Some transforms behave differently during inference than during training. For example, when we tokenize_messages, during inference we want add_end_tokens=False so that the model can "respond" to the prompt before end tokens are added. This flag allows for transforms to modify behavior depending on the situation.

#### Changelog
 - Added inference=False as a new parameter to every tokenizer/transform/template __call__
 - Updated tokenize_messages to `add_end_tokens= not inference` for every tokenizer
 - Updated VisionCrossAttentionMask to match kv_cache size if inference=True.
 
#### Test plan

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [ ] I have added an example to docs or docstrings;

### CODE REVIEW

**Review by joecummings — COMMENTED** [review]
tests?

File: torchtune/data/_prompt_templates.py [hunk 1]
```diff
@@ -87,13 +89,16 @@ def __init__(
     ):
         self.template = template
 
-    def __call__(self, messages: List[Message]) -> List[Message]:
+    def __call__(
+        self, messages: List[Message], inference: bool = False
```
Comment (RdoubleA) [inline torchtune/data/_prompt_templates.py:93]: this isn't being used anywhere?

File: torchtune/models/clip/_transform.py [hunk 2]
```diff
@@ -144,13 +144,16 @@ def __init__(
         self.tile_crop = tile_crop
         self.pad_tile_size = max_num_tiles if pad_max_tiles else None
 
-    def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
+    def __call__(
+        self, sample: Mapping[str, Any], inference: bool = False
```
Comment (RdoubleA) [inline torchtune/models/clip/_transform.py:148]: same here

File: torchtune/data/_prompt_templates.py [hunk 3]
```diff
@@ -25,13 +25,15 @@ class PromptTemplateInterface(Protocol):
...
         messages: List[Message],
+        inference: bool = False,
     ) -> List[Message]:
         """
         Format each role's message(s) according to the prompt template
 
         Args:
             messages (List[Message]): a single conversation, structured as a list
                 of :class:`~torchtune.data.Message` objects
+            inference (bool): Whether the template is being used for inference or not.
```
Comment (RdoubleA) [inline torchtune/data/_prompt_templates.py:36]: should describe what this practically changes

File: torchtune/modules/transforms/_transforms.py [hunk 4]
```diff
@@ -159,8 +176,10 @@ def __call__(self, sample: Mapping[str, Any]) -> Mapping[str, Any]:
             # Mask will be block of 1s at the corresponding interval in the text.
             # It is not a causal block because all the image tokens correspond
             # to a single image, so text tokens attend to all the image's tokens
-            mask = torch.zeros(text_seq_len, image_seq_len, dtype=torch.bool)
-            mask[start:end, :] = True
+            text_size = text_seq_len if not inference else self.max_seq_len
```
Comment (RdoubleA) [inline torchtune/modules/transforms/_transforms.py:179]: need to check if max_seq_len is not None

**Comment by RdoubleA** [comment]
should update respective tests with inference=True to make sure we get expected behavior

File: torchtune/data/_prompt_templates.py [hunk 5]
```diff
@@ -151,13 +156,15 @@ class ChatMLTemplate(PromptTemplateInterface):
...
         messages: List[Message],
+        inference: bool = False,
     ) -> List[Message]:
         """
         Format user, assistant, and system messages with appropriate tags.
 
         Args:
             messages (List[Message]): a single conversation, structured as a list
                 of `Message` objects
+            inference (bool): Whether the template is being used for inference or not.
```
Comment (RdoubleA) [inline torchtune/data/_prompt_templates.py:167]: probably not needed for this PR, but this is a good opportunity to use the inference flag for the empty assistant message logic below

**Comment by pbontrager** [comment]
I updated all of the current tests to tokenizer/transform `__call__` method

**Review by RdoubleA — APPROVED** [review]
